File: 3491-find-the-maximum-length-of-valid-subsequence-ii/3491-find-the-maximum-length-of-valid-subsequence-ii.py
class Solution:
    def maximumLength(self, nums: List[int], k: int) -> int:
        res = 0
        dp = [[0 for j in range(k)] for i in range(k)]
        for num in nums:
            mod = num % k
            for rem in range(k):
                diff = rem - mod + k % k
                dp[mod][rem] = dp[diff][rem] + 1
                res = max(res, dp[mod][rem])
        return res

File: 3491-find-the-maximum-length-of-valid-subsequence-ii/3491-find-the-maximum-length-of-valid-subsequence-ii.cpp
class Solution {
public:
    int maximumLength(vector<int>& nums, int k) {
        int res = 0;
        vector<vector<int>> dp(k, vector<int>(k, 0));
        for(int i = 0; i < nums.size(); i++)
        {
            int mod = nums[i] % k;
            for(int j = 0; j < k; j++)
            {
                int diff = (j - mod + k) % k;
                dp[mod][j] = dp[diff][j] + 1;
                res = max(res, dp[mod][j]);
            }
        }

        return res;
    }
};